addon/utils/disposable.js

```diff
@@ -0,0 +1,89 @@
+import Ember from 'ember';
+import { assert } from '@ember/debug';
+import { WILL_DESTROY_PATCHED } from '../utils/flags';
+
+const { WeakMap } = Ember;
+
+/**
+ * A map of instances/array of disposables. Only exported for
+ * testing purposes.
+ *
+ * @public
+ */
+export let registeredDisposables = new WeakMap();
+
+/**
+ * Registers a new disposable function to run for an instance. Will
+ * handle lazily creating a new array to store the disposables per
+ * instance if one does not exist. Will additionally patch an object's
+ * `willDestroy` hook to ensure the destroyables are run/destroyed when
+ * the object is destroyed.
+ *
+ * @public
+ * @param {*} obj the instance to store the disposable for
+ * @param {*} dispose a function that disposes of instance resources
+ * @returns a disposable object, which wraps the passed in `dispose` function
+ */
+export function registerDisposable(obj, dispose) {
+  assert(
+    'Called `registerDisposable` where `obj` is not an object',
+    typeof obj === 'object'
+  );
+  assert(
+    'Called `registerDisposable` where `dispose` is not a function',
+    typeof dispose === 'function'
+  );
+  assert(
+    'Called `registerDisposable` without implementing `willDestroy` that calls `runDisposables`',
+    !!obj[WILL_DESTROY_PATCHED]
+  );
+
+  let disposables = getRegisteredDisposables(obj);
+  let disposable = _toDisposable(dispose);
+
+  disposables.push(disposable);
+
+  return disposable;
+}
+
+/**
+ * Runs all disposables for a given instance.
+ *
+ * @public
+ * @param {*} obj the instance to run the disposables for
+ */
+export function runDisposables(obj) {
+  let disposables = registeredDisposables.get(obj);
+
+  if (!disposables) {
+    return;
+  }
+
+  for (let i = 0, l = disposables.length; i < l; i++) {
+    let disposable = disposables.pop();
+
+    disposable.dispose();
+  }
+}
+
+function getRegisteredDisposables(obj) {
+  let arr = registeredDisposables.get(obj);
+
+  if (!arr) {
+    registeredDisposables.set(obj, (arr = []));
+  }
+
+  return arr;
+}
+
+function _toDisposable(doDispose) {
+  return {
+    dispose() {
```

Comment:
this wrapper object doesn't seem to provide much value, combined with my above suggestions for iteration change. We could simply store `doDispose` in the array directly, no need to allocate more stuff.

Reply:
I agree, and had been thinking about discussing with @rwjblue as to it's refactoring/simplifying.

Reply:
Agree. I've simplified, with an intent to make this API private in a backwards incompatible breaking change. Ultimately, this will help us converge towards whatever 'cleanup' mechanism is implemented in Ember itself, which will allow us to change the underlying implementation of `registerDisposable` to use that mechanism instead.